vllm/lora/ops/xla_ops/pallas.py

```diff
@@ -10,124 +12,487 @@
+    # laning
+    N_LORA_LANES = math.ceil(DIM_BLOCK / D)
+    if D != DI and N_LORA_LANES > 1 and N > 1:
+        pad_N = next_multiple_of(N, N_LORA_LANES) - N
+        new_N = N + pad_N
+
+        loras = torch.nn.functional.pad(loras, (0, 0, 0, 0, 0, pad_N))
+        loras = loras.reshape((new_N // N_LORA_LANES, DIM_BLOCK, L))
+        idxs = idxs // N_LORA_LANES
+        N, D, L = loras.shape
 
     # Pad the loras' rank if it's too low. This is to allow it to fit in a TPU
     # register. This has to happen in pytorch, doing it in Jax will lead to NaNs
-    L1 = L
-    if LORA_RANK_BLOCK > L or L % LORA_RANK_BLOCK != 0:
-        L1 = (L // LORA_RANK_BLOCK + 1) * LORA_RANK_BLOCK
+    pad_L = 0
+    if LORA_BLOCK > L or L % LORA_BLOCK != 0:
+        pad_L = next_multiple_of(L, LORA_BLOCK) - L
 
-    D1 = D
-    if DIM_BLOCK_SIZE > D or D % DIM_BLOCK_SIZE != 0:
-        D1 = (D // DIM_BLOCK_SIZE + 1) * DIM_BLOCK_SIZE
+    pad_D = 0
+    if DIM_BLOCK > D or D % DIM_BLOCK != 0:
+        pad_D = next_multiple_of(D, DIM_BLOCK) - D
 
-    T1 = T
-    if TOKENS_BLOCK > T or T % TOKENS_BLOCK != 0:
-        T1 = (T // TOKENS_BLOCK + 1) * TOKENS_BLOCK
+    pad_T = 0
+    if TOKEN_BLOCK > T or T % TOKEN_BLOCK != 0:
+        pad_T = next_multiple_of(T, TOKEN_BLOCK) - T
 
-    if D1 != D or L1 != L:
-        loras = torch.nn.functional.pad(loras, (0, D1 - D, 0, L1 - L, 0, 0))
-    if D1 != D or T1 != T:
-        inputs = torch.nn.functional.pad(inputs, (0, D1 - D, 0, T1 - T))
-        if T1 != T:
-            idxs = torch.nn.functional.pad(idxs, ((0, T1 - T)))
+    if pad_D != 0 or pad_L != 0:
+        loras = torch.nn.functional.pad(loras, (0, pad_L, 0, pad_D, 0, 0))
+    if pad_D != 0 or pad_T != 0:
+        inputs = torch.nn.functional.pad(inputs, (0, pad_D, 0, pad_T))
+        if pad_T != T:
+            idxs = torch.nn.functional.pad(idxs, ((0, pad_T)))
+
+    jax_import_guard()
+
+    kernel = make_kernel_from_pallas(
+        functools.partial(_bgmv_expand,
+                          TOKEN_BLOCK=TOKEN_BLOCK,
+                          LORA_BLOCK=LORA_BLOCK,
+                          DIM_BLOCK=DIM_BLOCK), bgmv_expand_shape_function)
 
     return kernel(idxs, inputs, loras)[:T, :L]
 
 
-@impl(XLA_LIB, "bgmv", "CompositeExplicitAutograd")
-def bgmv_non_xla(inputs: torch.Tensor, loras: torch.Tensor,
-                 idxs: torch.IntTensor):
+@impl(XLA_LIB, "bgmv_expand", "CompositeExplicitAutograd")
+def bgmv_expand_non_xla(inputs: torch.Tensor, loras: torch.Tensor,
+                        idxs: torch.IntTensor):
     T, _ = inputs.shape
 
     if len(loras.shape) == 4:
         loras = loras.squeeze(axis=1)
 
-    _, L, _ = loras.shape
+    _, _, L = loras.shape
 
     return torch.empty((T, L), device=inputs.device)
+
+
+def largest_divisor(n: int, divs: List[int]) -> int:
+    for div in sorted(divs, reverse=True):
+        if n % div == 0:
+            return div
+    return max(divs)
+
+
+def next_multiple_of(n: int, mult: int) -> int:
```

Comment:
nit: are these functions not in any utils file already? I swear I've seen them pop up multiple times throughout the codebase.

Reply:
I can't seem to find them in any utils files